### ISSUE TITLE
ocrd workspace bag: downloaded files are overwritten

### ISSUE THREAD

**Issue by MehmedGIT** [OCR-D/core#1129]
The download of remote images is currently broken with core v2.58.1. Consider the following part of a mets file.

```xml
<mets:file ID="FILE_0001_DEFAULT" MIMETYPE="image/jpg">
  <mets:FLocat xmlns:xlink="http://www.w3.org/1999/xlink" LOCTYPE="URL" xlink:href="https://content.staatsbibliothek-berlin.de/dc/PPN631277528-00000001/full/max/0/default.jpg"/>
<mets:FLocat xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="DEFAULT/FILE_0001_DEFAULT.jpg" LOCTYPE="OTHER" OTHERLOCTYPE="FILE"/></mets:file>
<mets:file ID="FILE_0002_DEFAULT" MIMETYPE="image/jpg">
  <mets:FLocat xmlns:xlink="http://www.w3.org/1999/xlink" LOCTYPE="URL" xlink:href="https://content.staatsbibliothek-berlin.de/dc/PPN631277528-00000002/full/max/0/default.jpg"/>
<mets:FLocat xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="DEFAULT/FILE_0002_DEFAULT.jpg" LOCTYPE="OTHER" OTHERLOCTYPE="FILE"/></mets:file>
<mets:file ID="FILE_0003_DEFAULT" MIMETYPE="image/jpg">
  <mets:FLocat xmlns:xlink="http://www.w3.org/1999/xlink" LOCTYPE="URL" xlink:href="https://content.staatsbibliothek-berlin.de/dc/PPN631277528-00000003/full/max/0/default.jpg"/>
<mets:FLocat xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="DEFAULT/FILE_0003_DEFAULT.jpg" LOCTYPE="OTHER" OTHERLOCTYPE="FILE"/></mets:file>
<mets:file ID="FILE_0004_DEFAULT" MIMETYPE="image/jpg">
```

After downloading the images, there is only a single image (of the last page) named `default.jpg`. It seems the wrong `<mets:FLocat>` is used to parse the name of the images to be downloaded.

I have not investigated in depth yet but I think https://github.com/OCR-D/core/pull/1113 broke it.

**Comment by kba** [comment]
This might be because of bad defaults for the filename - can you share the workspace so I can debug further?

**Comment by MehmedGIT** [comment]
> This might be because of bad defaults for the filename - can you share the workspace so I can debug further?

This is the workspace: https://content.staatsbibliothek-berlin.de/dc/PPN631277528.mets.xml 

It breaks the Operandi endpoint that creates a workspace from mets URL. The `DEFAULT` filegroup is used by default.

**Comment by kba** [comment]
Strange, if I do

```
ocrd workspace clone 'https://content.staatsbibliothek-berlin.de/dc/PPN631277528.mets.xml'
ocrd workspace find -G DEFAULT --download
```

then the `DEFAULT` folder contains, as expected:

```
❯ ls DEFAULT 
FILE_0001_DEFAULT.jpg  FILE_0006_DEFAULT.jpg  FILE_0011_DEFAULT.jpg  FILE_0016_DEFAULT.jpg  FILE_0021_DEFAULT.jpg  FILE_0026_DEFAULT.jpg
FILE_0002_DEFAULT.jpg  FILE_0007_DEFAULT.jpg  FILE_0012_DEFAULT.jpg  FILE_0017_DEFAULT.jpg  FILE_0022_DEFAULT.jpg  FILE_0027_DEFAULT.jpg
FILE_0003_DEFAULT.jpg  FILE_0008_DEFAULT.jpg  FILE_0013_DEFAULT.jpg  FILE_0018_DEFAULT.jpg  FILE_0023_DEFAULT.jpg  FILE_0028_DEFAULT.jpg
FILE_0004_DEFAULT.jpg  FILE_0009_DEFAULT.jpg  FILE_0014_DEFAULT.jpg  FILE_0019_DEFAULT.jpg  FILE_0024_DEFAULT.jpg  FILE_0029_DEFAULT.jpg
FILE_0005_DEFAULT.jpg  FILE_0010_DEFAULT.jpg  FILE_0015_DEFAULT.jpg  FILE_0020_DEFAULT.jpg  FILE_0025_DEFAULT.jpg
```

The relevant part of the `download_file` method in `OcrdWorkspace`:

```python
# If f.url is set, download the file to the workspace                                                              
basename = '%s%s' % (f.ID, MIME_TO_EXT.get(f.mimetype, '')) if f.ID else f.basename                                
f.local_filename = self.resolver.download_to_directory(self.directory, f.url, subdir=f.fileGrp, basename=basename) 
```

I.e. the basename should default to `{{ ID }}{{ extension }}` (which it does for my call) and default to the file's `basename` (which seems to be the case for your call) if `ID is not set.

So, for some reason, the `mets:file/@ID` seems unset. Do you have an idea why?

**Comment by MehmedGIT** [comment]
With the CLI it works for me too. The problem is there when trying to retrieve files programmatically.

```shell
mm@MM-Notebook:~/Desktop/ocrd_network_files/example_ws6$ ocrd workspace clone 'https://content.staatsbibliothek-berlin.de/dc/PPN631277528.mets.xml'
10:05:34.091 INFO ocrd.models.utils.is_oai_content - response data root.tag: '{http://www.loc.gov/METS/}mets'
/home/mm/Desktop/ocrd_network_files/example_ws6
mm@MM-Notebook:~/Desktop/ocrd_network_files/example_ws6$ ocrd workspace find -G DEFAULT --download
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
```

```shell
mm@MM-Notebook:~/Desktop/ocrd_network_files/example_ws6/DEFAULT$ ls
FILE_0001_DEFAULT.jpg  FILE_0006_DEFAULT.jpg  FILE_0011_DEFAULT.jpg  FILE_0016_DEFAULT.jpg  FILE_0021_DEFAULT.jpg  FILE_0026_DEFAULT.jpg
FILE_0002_DEFAULT.jpg  FILE_0007_DEFAULT.jpg  FILE_0012_DEFAULT.jpg  FILE_0017_DEFAULT.jpg  FILE_0022_DEFAULT.jpg  FILE_0027_DEFAULT.jpg
FILE_0003_DEFAULT.jpg  FILE_0008_DEFAULT.jpg  FILE_0013_DEFAULT.jpg  FILE_0018_DEFAULT.jpg  FILE_0023_DEFAULT.jpg  FILE_0028_DEFAULT.jpg
FILE_0004_DEFAULT.jpg  FILE_0009_DEFAULT.jpg  FILE_0014_DEFAULT.jpg  FILE_0019_DEFAULT.jpg  FILE_0024_DEFAULT.jpg  FILE_0029_DEFAULT.jpg
FILE_0005_DEFAULT.jpg  FILE_0010_DEFAULT.jpg  FILE_0015_DEFAULT.jpg  FILE_0020_DEFAULT.jpg  FILE_0025_DEFAULT.jpg
```

**Comment by MehmedGIT** [comment]
You can reproduce the issue with the following code snippet based on Operandi:
```python
from ocrd import Resolver
from ocrd.workspace import Workspace
from ocrd.workspace_bagger import WorkspaceBagger
from typing import List

DEFAULT_METS_BASENAME = "mets.xml"
DEFAULT_FILE_GRP = "DEFAULT"

def create_ws(
    mets_url: str,
    workspace_id: str,
    bag_dest: str,
    mets_basename: str = DEFAULT_METS_BASENAME,
    preserve_file_grps: List[str] = None
) -> str:
    if not preserve_file_grps:
        preserve_file_grps = [DEFAULT_FILE_GRP]

    resolver = Resolver()
    # Create an OCR-D Workspace from a remote mets URL
    # without downloading the files referenced in the mets file
    workspace = resolver.workspace_from_url(
        mets_url=mets_url,
        clobber_mets=False,
        mets_basename=mets_basename,
        download=False
    )

    remove_groups = [x for x in workspace.mets.file_groups if x not in preserve_file_grps]
    for group in remove_groups:
        workspace.remove_file_group(group, recursive=True, force=True)
    workspace.save_mets()

    # Resolves and downloads all file groups and files in the mets file
    WorkspaceBagger(resolver).bag(
        workspace,
        dest=bag_dest,
        ocrd_identifier=workspace_id,
        processes=1   # Do not pass higher processes number
    )
    return workspace.directory
    
ws_dir = create_ws(mets_url="https://content.staatsbibliothek-berlin.de/dc/PPN631277528.mets.xml", workspace_id="test123", bag_dest='/tmp/ocrd_ws_temp123')
```

I have a better idea of why it is failing. The `download=False` is causing the issue. I put that to False because by default all file groups were downloaded and this was not the desired behavior. If I revert that to True, some file group URLs are local and not downloadable, e.g.:

```shell
Traceback (most recent call last):
  File "download_ws.py", line 46, in <module>
    ws_dir = create_ws(mets_url="https://content.staatsbibliothek-berlin.de/dc/PPN631277528.mets.xml", workspace_id="test123", bag_dest='/tmp/ocrd_ws_temp123')
  File "download_ws.py", line 24, in create_ws
    workspace = resolver.workspace_from_url(
  File "/home/mm/venv-operandi/lib/python3.8/site-packages/ocrd/resolver.py", line 221, in workspace_from_url
    workspace.download_file(f)
  File "/home/mm/venv-operandi/lib/python3.8/site-packages/ocrd/workspace.py", line 203, in download_file
    f.local_filename = self.resolver.download_to_directory(self.directory, f.url, subdir=f.fileGrp, basename=basename)
  File "/home/mm/venv-operandi/lib/python3.8/site-packages/ocrd/resolver.py", line 92, in download_to_directory
    raise FileNotFoundError(f"File path passed as 'url' to download_to_directory does not exist: '{url}")
FileNotFoundError: File path passed as 'url' to download_to_directory does not exist: 'file:///goobi/tiff001/sbb/PPN631277528/00000001.tif
```

Potentially, this should still be fixed in the core to not try downloading files with the prefix `file:///` which is typical for Goobi.

The second reason I set the flag to false was that downloading all file groups is not efficient when it comes to transferring the ocrd zips around. Maybe this should be more flexible and allow downloading only specified file groups when creating the zip. However, I understand that this would then cause some hash match problems. Hence, I remove the file groups that are not needed before creating the zip. This is also not ideal, because the removed information is then lost forever.

**Comment by MehmedGIT** [comment]
Btw, I have fixed the issue by simply adding the manual download after `workspace.save_mets()`:
```python
ws_files = workspace.find_files()
for ws_file in ws_files:
    workspace.download_file(f=ws_file)
```

**Comment by kba** [comment]
Thanks for the explanation. The problem is very likely the WorkspaceBagger, I remember that we changed the logic for the filenames there a while ago, I will investigate and get back to you.

But I think your workaround is what I actually would do, be explicit about what you want to download instead of relying on the download options of either `workspace_from_url(..., download=True)` or downloads as the side effect of bagging.

**Comment by kba** [comment]
About the `file://` URL: When we started developing core, the assumption was that we will work only with proper URL, including `file://` URL for the downloaded files. When that proved to be very inconvenient, we pivoted to relative file paths instead of URL.

With Kitodo/Goobi there is also the problem of Java-style `file:/path/to` URL without the double slash `//`.

We could just warn about those instead of raising FNFE.
